freeglut/freeglut/src/fg_callback_macros.h

```diff
@@ -0,0 +1,412 @@
+ * Info:
+ *
+ * This took a while to figure out, so be sure try to understand what is happening so that you can ensure that whatever you
+ * change won't break other areas.
+ * 
+ * If you are just adding a new callback/changing it's argument count, just go to the bottom of the file.
+ *
+ * This whole file exists purely for the sake of preventing the need to implement additional parsing logic for each callback
+ * to pass user arguments. Of course, the necessity to support older compilers means that, as seen in the line above, there
+ * is still a requirement to add/modify code to handle callbacks. If freeglut ever requires newer compilers (at minimum, ones
+ * that support C99 or higher), code can very slowly be removed from this file. Even better would be if the C standard eventually
+ * supports something similar to what GCC has implemented or offers an alternative. Another option is if C++ would be "allowed" by
+ * project maintaners, as then templates can be used and function overloading. Ironically, the template would probably look worse
+ * then the GCC macro, so maybe it's good to stay as is.
+ *
+ * Onto the different "versions" of macros:
+ * 
+ * There is one for GCC/Clang(/and supposedly the Intel compiler) which supports the non-standard ##__VA_ARGS__ token. The code may 
+ * look ugly, but the result is, if this was standard, no one would ever need to open this file unless they were curious (or needed 
+ * more then 5 arguments for a callback, but that's trivial to add). It works by adding many fake macros to a "picker" macro
+ * (PP_HAS_ARGS_IMPL2) which then indictaes which macro counter to use. As we can already use variadic macros (the VA in __VA_ARGS__),
+ * this just becomes a "reuse the arguments*.
+ * 
+ * The next is for any non-GCC/Clang/etc. compiler *cough* MSVC/compiler you probably shouldn't be using *cough* that supports C99
+ * by default. It requires each callback to have a specific argument count passthrough macro. The only reason there are specific
+ * count macros is so that (see paraghraph below) don't need have their own set of callback macros. Ideally, there would only be
+ * ZERO and ONE_OR_MORE. This works by having callback-specific macros call a specific handler macro to return user data (ZERO) or
+ * return one or more arguments along with userData (ONE_OR_MORE) where, with variadic macros, it just reuses the arguments.
+ *
+ * The last set is for the poor individual who has to use a compiler that doesn't support C99 by default, or may not support it at
+ * all. Stuff like MSVC6... It works by having a specific-count macro that "extracts" each argument to have them reused without the
+ * parathesis.
+ *
+ * A note on parathesis, as earlier mentioned, if the GCC variadic macro element was standard, then instead of needing:
+ *
+ * func EXPAND_WCB(Mouse)(( (GLUT_LEFT_BUTTON, GLUT_DOWN, 10, 30), userData));
+ *
+ * ...you can do the following:
+ *
+ * func EXPAND_WCB (GLUT_LEFT_BUTTON, GLUT_DOWN, 10, 30);
```

Comment:
And... screwed up both comments. So, since GCC variadic macro extension is not standard, a workaround must be written. This is explaining how the code would look different if it was a standard.